doc/tutorial.md

```diff
@@ -0,0 +1,594 @@
+
+```bash
+service restart zebra
+service restart ospfd
+```
+
+Hopefully (if `OSPF` works), the VIP is accessible by client:
+
+```bash
+client: curl 123.1.2.3
+```
+
+<a id='fnat-keepalive'/>
+
+## Full-NAT with Keepalived (one-arm)
+
+This is an example for FullNAT used in internal network (LAN). `Keepalived`  (*`DPVS` modified version*) is used for to make DPVS works as *Master/Backup* model.
+
+![fnat-keepalive](pics/fnat-keepalived.png)
+
+By using `keepalived`, routes, `LIP`, `VIP` and `RS` can be configured through `keepalived` config file. **Note** the configure parameters for `DPVS` modified `keepalived` is slight **different** from original `keepalived`.
+
+```
+$ cat /etc/keepalived/keepalived.conf
+! Configuration File for keepalived
+
+global_defs {
+    notification_email {
+        foo@example.com
+    }
+    notification_email_from bar@example.com
+    smtp_server 1.2.3.4
+    smtp_connect_timeout 60
+    router_id DPVS_DEVEL
+}
+
+local_address_group laddr_g1 {
+    192.168.100.200 dpdk0    # use DPDK interface
+    192.168.100.201 dpdk0    # use DPDK interface
+}
+
+#
+# VRRP section
+#
+vrrp_instance VI_1 {
+    state MASTER                  # master
+    interface dpdk0.kni           # should be kni interface
+    dpdk_interface dpdk0          # should be DPDK interface
+    virtual_router_id 123         # VID should be unique in network
+    priority 100                  # master's priority is bigger than worker
+    advert_int 1
+    authentication {
+        auth_type PASS
+        auth_pass ****
+    }
+
+    virtual_ipaddress {
+        192.168.100.254
+    }
+}
+
+#
+# Virtual Server Section
+#
+virtual_server_group 192.168.100.254-80 {
+    192.168.100.254 80
+}
+
+virtual_server group 192.168.100.254-80 {
+    delay_loop 3
+    lb_algo rr         # scheduling algorithm Round-Robin
+    lb_kind FNAT       # Forwarding Mode Full-NAT
+    protocol TCP       # Protocol TCP
+
+    laddr_group_name laddr_g1   # Local IP group-ID
+
+    real_server 192.168.100.2 80 { # real-server
+        weight 100
+        inhibit_on_failure
+        TCP_CHECK {    # health check
+            nb_sock_retry 2
+            connect_timeout 3
+            connect_port 80
+        }
+    }
+
+    real_server 192.168.100.3 80 { # real-server
+        weight 100
+        inhibit_on_failure
+        TCP_CHECK { # health check
+            nb_sock_retry 2
+            connect_timeout 3
+            connect_port 80
+        }
+    }
+}
+```
+
+The keepalived config for backup is same with Master, except the `state` should be 'BACKUP', and `priority` should be lower.
+
+```
+vrrp_instance VI_1 {
+    state BACKUP
+    priority 80
+    ... ...
+}
+```
+
+Start `keepalived` on both Master and Backup.
+
+```bash
+./keepalived -f /etc/keepalived/keepalived.conf
+```
+
+For **test only**, add `VIP` and *routes* to DPDK interface manually on Master (do not set VIP on both master and backup), they should be added to keepalived configure file.
```

Comment:
vip can be added to dpdk interfaces automatically, do not need manual configuration